Create detailed custom-target-porting.md

maclobdell opened the pull request:
Add tutorial about porting Mbed OS to a custom board.

This replaces https://github.com/ARMmbed/mbed-os-5-docs/pull/1130.

Due to difficulty rebasing the previous PR, I went ahead and created a new one.  

I rewrote the document based on feedback received.  My goal was to provide more clarity and helpful details.   

There might be formatting issues.  Pleases review and let me know what I should fix.

Commit:
add new tutorial, more detailed than previous attempts

maclobdell commented:
@iriark01 for review

Commit:
Begin edits on custom-target-porting.md

Push up initial copy edits.

iriark01 commented:
@arekzaluski and @sophewilliams  I understand you did a lot of work on this topic - would you like to review this?

AnotherButler commented:
ping @arekzaluski @sophewilliams

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+
+When designing a custom board based on an existing Mbed Enabled board, you may need to make configuration and software changes to adapt Mbed OS to run on your new board. This is called a software port. The necessary changes may include adaptations for the unique design choices you have made for your new board, such as clocking, pin connections and peripheral usage. This can often be accomplished  ...
+
+Mbed OS supports target inheritance, which allows you to extend an existing microcontroller (MCU) target and just add software and configurations required for your board. You have the ability to add a file named `custom_target.json` to your project, which can store your custom target configurations without the need to modify `targets.json`.
```
I'm not clear on the break and order of information in the first two paragraphs.

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+
+When designing a custom board based on an existing Mbed Enabled board, you may need to make configuration and software changes to adapt Mbed OS to run on your new board. This is called a software port. The necessary changes may include adaptations for the unique design choices you have made for your new board, such as clocking, pin connections and peripheral usage. This can often be accomplished  ...
+
+Mbed OS supports target inheritance, which allows you to extend an existing microcontroller (MCU) target and just add software and configurations required for your board. You have the ability to add a file named `custom_target.json` to your project, which can store your custom target configurations without the need to modify `targets.json`.
```
```suggestion
Mbed OS supports target inheritance, which allows you to extend an existing microcontroller (MCU) target and just add software and configurations required for your board. You can add a file named `custom_target.json` to your project, which can store your custom target configurations without the need to modify `targets.json`.
```

iriark01 replied:
Note that this sentence assumes I know what targets.json is; might be nice to link to information about it, or any other information we want people to have before they try this tutorial (the assumption may be correct for our readers - I'm just pointing it out)

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+
+Mbed OS supports target inheritance, which allows you to extend an existing microcontroller (MCU) target and just add software and configurations required for your board. You have the ability to add a file named `custom_target.json` to your project, which can store your custom target configurations without the need to modify `targets.json`.
+
+This tutorial provides a quick overview on how to create a custom port using `custom_target.json`. It covers the most common methods. For detailed information covering all the ways you can configure targets, go to [adding and configuring targets](../reference/adding-and-configuring-targets.html).
```
```suggestion
This tutorial covers the most common methods used to create a custom port with `custom_target.json`. For detailed information covering all the ways you can configure targets, go to [adding and configuring targets](../reference/adding-and-configuring-targets.html).
```

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+
+## Extending an existing MCU target configuration
+
+As an example, consider a situation in which you are creating a new board based on an existing Mbed Enabled board. For this tutorial, consider a new board called `ImaginaryBoard` that is based on [DISCO-L475VG-IOT01A](https://os.mbed.com/platforms/ST-Discovery-L475E-IOT01A/).
```
Is that an example, or just the premise of this whole document?

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+
+1. Create a new file named `custom_targets.json` at the same level as the `mbed-os` directory:
+
+   Inspect the contents of `mbed-os/targets/targets.json`. For this example, search for `DISCO_L475VG_IOT01A`.
```
This feels like a new step.

iriark01 commented on docs/porting/custom-target-porting.md:
````diff
@@ -0,0 +1,259 @@
+   mbed new --program mbed-os-imaginary-port
+   ```
+  
+   This command imports `mbed-os` from the [official Mbed OS source repository](https://github.com/armmbed/mbed-os) into a new directory called `mbed-os-imaginary-port`.
````
If this is the point of this step, then it should be the first sentence in it. Something like "Import mbed-os by creating a new programme with Mbed CLI"

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+
+As an example, consider a situation in which you are creating a new board based on an existing Mbed Enabled board. For this tutorial, consider a new board called `ImaginaryBoard` that is based on [DISCO-L475VG-IOT01A](https://os.mbed.com/platforms/ST-Discovery-L475E-IOT01A/).
+
+`ImaginaryBoard` shares most of the same features of DISCO-L475VG-IOT01A, but it does not use `AnalogOut`, `AnalogIn`, `CAN` or `USB`. Some pins are also connected differently on the new board.
```
```suggestion
`ImaginaryBoard` shares most of the features of DISCO-L475VG-IOT01A, but it does not use `AnalogOut`, `AnalogIn`, `CAN` or `USB`. Some pins are also connected differently on the new board.
```

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+
+`ImaginaryBoard` shares most of the same features of DISCO-L475VG-IOT01A, but it does not use `AnalogOut`, `AnalogIn`, `CAN` or `USB`. Some pins are also connected differently on the new board.
+
+Assuming you have Mbed CLI installed, follow these steps to make the port.
```
Don't assume; link me to the installation instructions

iriark01 replied:
Do we make a port, or port?

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+  
+   This command imports `mbed-os` from the [official Mbed OS source repository](https://github.com/armmbed/mbed-os) into a new directory called `mbed-os-imaginary-port`.
+
+   Now, change directories into your new project:  
```
Whenever you use a time word, it's a good indication that you may need a numbered step.

iriark01 commented on docs/porting/custom-target-porting.md:
````diff
@@ -0,0 +1,259 @@
+  cd mbed-os-imaginary-port
+  ```
+
+1. Create a new file named `custom_targets.json` at the same level as the `mbed-os` directory:
````
What does : refer to?

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+
+   Inspect the contents of `mbed-os/targets/targets.json`. For this example, search for `DISCO_L475VG_IOT01A`.
+
+   Copy the contents from the `DISCO_L475VG_IOT01A` section into your `custom_targets.json` file. Be sure to include brackets `{ }` surrounding the content.
```
New step

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+
+   Copy the contents from the `DISCO_L475VG_IOT01A` section into your `custom_targets.json` file. Be sure to include brackets `{ }` surrounding the content.
+
+   Then make changes for your board. For example, after making changes, the full contents look like this:
```
New step

iriark01 replied:
Actually, you can consider having all the previous steps as a sort of "getting ready" bit, and then the interesting bits that follow them (where you actually change the target file) are just a whole new section, with a nice header indicating that

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+
+   - The board name was changed from `DISCO_L475VG_IOT01A` to `IMAGINARYBOARD` so the board can be uniquely identified.
+
+   - The `detect_code` was changed from `0764` to `1234`. The `detect_code` is a unique number that identifies the board to the Mbed OS test tools. This number is typically built into the debug interface.
```
"Typically built" does that mean that's where I get it from? What if it's not there - how do I know which number to define?

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+
+   - The `macros_add` section was changed to remove `USBHOST_OTHER` because USB is not used on the new board.
+
+   - The `device_has_add` section was changed to not add the `ANALOGOUT`, `CAN`, `USBDEVICE` drivers because those features are not used on the new board.
```
```suggestion
   - The `device_has_add` section was changed to not add the `ANALOGOUT`, `CAN`, and `USBDEVICE` drivers because those features are not used on the new board.
```

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+
+   - The `device_has_add` section was changed to not add the `ANALOGOUT`, `CAN`, `USBDEVICE` drivers because those features are not used on the new board.
+
+   - A new section, `device_has_remove`, was added. This removes `ANALOGIN`, `I2CSLAVE`, and `I2C_ASYNCH` drivers because these features are also not used.  The reason why `device_has_remove` is used in this case is because the board is inheriting from the MCU Family configuration `FAMILY_STM32`, which has those drivers added by default.
```
```suggestion
   - A new section, `device_has_remove`, was added. This removes the `ANALOGIN`, `I2CSLAVE`, and `I2C_ASYNCH` drivers because these features are also not used. The reason `device_has_remove` is used in this case is that the board is inheriting from the MCU Family configuration `FAMILY_STM32`, which has those drivers added by default.
```

iriark01 replied:
I feel like the point about having to remove inherited things can be made more prominent.

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+
+   All the other configurations for the board are inherited from the MCU Family configuration called `FAMILY_STM32`.  
+
+   Here are some other typical changes that might be needed.
```
"might be needed" already means "optional", so you don't need to open the bullets with that word

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+   
+   - (Optional) You can also use `device_has_add` to add additional drivers.
+
+   <span class="notes">**Note:** If you choose to add a driver, you may have to provide the driver implementation if it is not already available for your hardware.</span>
```
```suggestion
   <span class="notes">**Note:** If you choose to add a driver that is not already available for your hardware, you will have to provide the driver implementation.</span>
```

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+
+   <span class="notes">**Note:** If you choose to add a driver, you may have to provide the driver implementation if it is not already available for your hardware.</span>
+
+   - (Optional) Similarly, you can use `features_add`, `features_remove`, `components_add`, `components_remove`, `macros_add` and `macros_remove` to add/remove configurations.
```
```suggestion
   - (Optional) Similarly, you can use `features_add`, `features_remove`, `components_add`, `components_remove`, `macros_add` and `macros_remove` to add or remove configurations.
```

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+
+Boards typically inherit files that support the MCU, MCU family and MCU vendor. When adding a new board, you need to add a new set of files for the board.
+
+You may be wondering why there are more directory levels than target configuration levels. This is because many targets use the `extra_labels_add` feature in the target configuration. The following keywords `STM32L4`, `STM32L475xG`, `STM32L475VG` resolve to `TARGET_STM32L4`, `TARGET_STM32L475xG`, `TARGET_STM32L475VG` respectively. With those labels applied, these directory names get included in t ...
```
```suggestion
You may be wondering why there are more directory levels than target configuration levels. This is because many targets use the `extra_labels_add` feature in the target configuration. The keywords `STM32L4`, `STM32L475xG`, `STM32L475VG` resolve to `TARGET_STM32L4`, `TARGET_STM32L475xG`, `TARGET_STM32L475VG` respectively. With those labels applied, these directory names get included in the build for this target.
```

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+   
+   Inspect the files at `mbed-os\targets\TARGET_STM\TARGET_STM32L4\TARGET_STM32L475xG\TARGET_DISCO_L475VG_IOT01A`. You should find the following files or similar:
+   
+   `PeripheralNames.h`, `PeripheralPins.c`, `PinNames.h`, `system_clock.c`
```
There's a lot of repetition in this step. You can combine which files I should see with what they provide.

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+      
+1. Modify the files.
+
+   `PinNames.h` is the most common file to be edited. For our example, the ImaginaryBoard is using I2C but connected to different supported signals. Change the I2C pin macro definitions from:
```
```suggestion
   `PinNames.h` is the most common file to be edited. For our example, the ImaginaryBoard uses I2C connected to different supported signals. Change the I2C pin macro definitions from:
```

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+   - `PinNames.h` sets macros for pins that define their function.
+   - `system_clock.c` vendor specific file that initializes the system and sets up the clocks.
+      
+1. Modify the files.
```
Again, maybe have all the previous steps as a "getting ready" bit and this as a new header, since it's the meat of the issue

iriark01 commented on docs/porting/custom-target-porting.md:
````diff
@@ -0,0 +1,259 @@
+   I2C_SDA     = PC_1,
+   ```
+   
+   This edit is mainly for convenience if your application code uses standard I2C pin names `I2C_SDA` and `I2C_SCL`.
````
This feels like you could add some interesting information.

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+   
+   You may also choose to add or remove peripherals, add or remove pins or change the clock frequency by editing `PeripheralNames.h`, `PeripheralPins.c` or `system_clock.c`. In our example, for simplicity, we will not edit these files.
+   
+1. (Optional) Add additional files.
```
You don't need both lines of this step; the second line includes all the information of the first line.

iriark01 commented on docs/porting/custom-target-porting.md:
````diff
@@ -0,0 +1,259 @@
+    }
+    ``` 
+
+    This blinks an LED. Just make sure LED1 is defined for the board.
````
"Just make sure LED1 is defined" is maybe an important point that should be made more prominent.

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+    
+1. Compile the project.
+   
+  With the target configuration set and files added, it is ready to compile. Run the command:
```
Not sure this sentence adds new info.

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+
+1. Program the board.  
+
+  You can test this simple example using a `DISCO-L475VG-IOT01A`. If you actually created a `ImaginaryBoard` board, you could use that too.  
```
```suggestion
  You can test this simple example using a `DISCO-L475VG-IOT01A`. If you actually created an `ImaginaryBoard` board, you could use that too.  
```

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+
+  You can test this simple example using a `DISCO-L475VG-IOT01A`. If you actually created a `ImaginaryBoard` board, you could use that too.  
+  
+  <span class="notes">**Note:** Unless your board has an Mbed Enabled debug interface, you need a method of flashing the memory on your board. </span>
```
Should this come earlier?

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+  
+  <span class="notes">**Note:** Unless your board has an Mbed Enabled debug interface, you need a method of flashing the memory on your board. </span>
+  
+  Since the `DISCO-L475VG-IOT01A` has a Mbed Enabled debug interface (STLink in this case), we can use drag-and-drop programming to flash the board.
```
```suggestion
  Since the `DISCO-L475VG-IOT01A` has an Mbed Enabled debug interface (STLink in this case), we can use drag-and-drop programming to flash the board.
```

iriark01 replied:
Are we flipping a bit between "We" and "You"?

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+  
+  Locate the binary file and drag it onto the disk drive name for the board (for example, `DIS_L4IOT`).
+  
+  After the file transfer is complete, press the reset button on the board. You should see the LED blinking.
```
```suggestion
  When the file transfer is complete, press the reset button on the board. You should see the LED blinking.
```

iriark01 replied:
New step

iriark01 commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,259 @@
+  
+  After the file transfer is complete, press the reset button on the board. You should see the LED blinking.
+ 
+1. (Optional) Run automated tests.
```
Maybe new header?

iriark01 commented on docs/porting/custom-target-porting.md:
````diff
@@ -0,0 +1,259 @@
+  mbed test -m IMAGINARYBOARD -t <toolchain>
+  ```
+  
+  The tests should start running.
````
Can we link to some information about testing?

Commit:
Update docs/porting/custom-target-porting.md

Co-Authored-By: Irit Arkin <irit.arkin@arm.com>

Commit:
Update docs/porting/custom-target-porting.md

Co-Authored-By: Irit Arkin <irit.arkin@arm.com>

Commit:
Update docs/porting/custom-target-porting.md

Co-Authored-By: Irit Arkin <irit.arkin@arm.com>

Commit:
Update docs/porting/custom-target-porting.md

Co-Authored-By: Irit Arkin <irit.arkin@arm.com>

Commit:
Update docs/porting/custom-target-porting.md

Co-Authored-By: Irit Arkin <irit.arkin@arm.com>

Commit:
Update docs/porting/custom-target-porting.md

Co-Authored-By: Irit Arkin <irit.arkin@arm.com>

Commit:
Update docs/porting/custom-target-porting.md

Co-Authored-By: Irit Arkin <irit.arkin@arm.com>

Commit:
Update docs/porting/custom-target-porting.md

Co-Authored-By: Irit Arkin <irit.arkin@arm.com>

Commit:
rewrites and edits to custom target porting

Commit:
Edit custom-target-porting.md

Edit file, mostly for active voice, consistent headings, use of numbered lists and to address comments.

AnotherButler commented:
@iriark01 Could you please review again?

AnotherButler commented:
@arekzaluski @sophewilliams Would y'all like to review?

bulislaw commented on docs/porting/custom-target-porting.md:
```diff
@@ -0,0 +1,258 @@
+#### Changes
+
+1. The board name changed from `DISCO_L475VG_IOT01A` to `IMAGINARYBOARD`, so the board can be uniquely identified.
+1. The `detect_code` changed from `0764` to `1234`. The `detect_code` is a unique number that identifies the board to the Mbed OS test tools. For Mbed Enabled boards, this number is exposed through the debug interface with Mbed CLI by typing `mbedls`.
```
Do we need some information how to set it? When do you need to get it from us and when it's ok to just pick some? How to avoid clashes? Consideration based on public vs private port? I guess it deserves a link to a separate doc.

maclobdell replied:
Looking through our documentation - I see it mentioned a few times, but never really see it explained well and the terminology is inconsistent.  

- Sometimes its called a **board id** (https://www.mbed.com/en/about-mbed/mbed-enabled/requirements/).  
- Sometimes it is called **four-digit hexadecimal identifier** (https://os.mbed.com/docs/mbed-os/v5.14/porting/mbed-ls.html)
- Sometimes its called a **Platform ID** (https://github.com/ARMmbed/mbed-os-tools/tree/master/packages/mbed-ls#adding-platform-support).  
- Sometimes its called **unique board ID** (https://os.mbed.com/docs/mbed-os/v5.14/porting/setting-up.html)
- Sometimes its called **detect_code** (https://github.com/ARMmbed/mbed-os/blob/master/targets/targets.json) 

I like Board ID, but Platform ID or the others seem fine.  

For me the key points are: 
1. Board IDs are unique identifiers assigned by the Arm Mbed team.
2. Board IDs are used by the Mbed tools to identify the connected board.
3. You really only need your own unique ID if you plan to push your platform to the Mbed community.
4. You can make Mbed CLI recognize your own custom board even without getting your own Board ID.
5. Contact a technical account manger or Mbed support if you would like to request a board ID.  

@AnotherButler - can you recommend where this information would be best located?

maclobdell replied:
I just added a new commit for this PR with changes related to the Platform ID. Also, I made pull request https://github.com/ARMmbed/mbed-os-5-docs/pull/1175 to align the terminology and add more detail.  Please let me know if this addresses the comments.

bulislaw reviewed (approved):
Looks great, I would just add some more details and considerations around the detect code.

AnotherButler commented:
@maclobdell Can you make these changes requested by @bulislaw ?

Commit:
add detect code clarifications

AnotherButler commented:
Thanks for the updates, @maclobdell 

This is failing Circle CI because of something unrelated. Could you please rebase, and then I'll merge?

AnotherButler commented:
Closing in favor of https://github.com/ARMmbed/mbed-os-5-docs/commit/6be8ae4aaf634ea8ebdf87c69b8c61044e5715fc
